### DISCUSSION TITLE
How to backup and restore certain tables only?

### DISCUSSION THREAD

**Discussion by mytc** [simolus3/drift#1253] [Q&A]
There is a discussion on how to restore and backup in flutter moor [here.](https://github.com/simolus3/moor/issues/376).
It works only for backup and restore entire database. 
But what I want is to export and restore specific tables only. 

Hence I wonder is there any simple way to do this?
I am thinking of using csv, but it seem complicated when the tables are related with each other. Eg, User table, and User entry where user entry have foreign key of user. 
Other option would be json file. 
And what about the performance of the process, if the table has, say more than 10k of entries?

**Comment by simolus3** [comment]
External representations like json or csv should work. When there's a lot of data, you may have to write it in small chunks.

However, if you just need backup/restore, I think the easiest way is to just use sqlite3 to copy the data over. I haven't tried it, but I think something like this should work:

```dart
extension CreateBackup on GeneratedDatabase {
  Future<void> createBackup(File to, List<TableInfo> tables) async {
    await customStatement('ATTACH DATABASE ? AS backup_target', [to.path]);

    final migrator = Migrator(this);
    for (final table in tables) {
      final sourceName = table.actualTableName;
      final targetName = 'backup_target.${table.actualTableName}';

      final aliased = alias(table, targetName) as TableInfo;
      await migrator.createTable(aliased);

      await customStatement(
          'INSERT INTO $targetName SELECT * FROM $sourceName');
    }

    await customStatement('DETACH backup_target');
  }
}
```

To restore a backup, you could once again attach the additional database and copy values over in the other direction.

**Reply by mytc** [reply]
thanks! Will give it a try.

**Reply by mytc** [reply]
I encountered few problem when using the code you provided.

```
 await customStatement('ATTACH DATABASE ? AS backup_target', [to.path]);
      final migrator = Migrator(this);
      final sourceName = 'users';
      final targetName = 'bckp_users';
      $UsersTable users = $UsersTable(this);
      final aliased = alias(users, targetName);
      await migrator.createTable(aliased);
      await customStatement(
          'INSERT INTO $targetName SELECT * FROM $sourceName');//here the exception occured.
      await customStatement('DETACH backup_target');
```
I got this 
`SqliteException (SqliteException(1555): UNIQUE constraint failed: bckp_users.id); `
Apparently the source table got a unique id which cannot directly inserted to target table.

Also, when second time I relaunch backup from a tap on a button in the app, it will stated that:
`SqliteException (SqliteException(1): database backup_target is already in use, SQL logic error (code 1))`
Is there anyway to check the if the backup database is already in use?

**Reply by simolus3** [reply]
I don't really understand how the constraint can fail in the target table without failing in the source table as well. You're also not creating the table in the backup if the table name doesn't start with `backup_target.`.

But on another thought, I think creating the table with an alias is rather fragile. I'd open another database instance for the backup file, create the tables in there (just opening it is enough to create tables, use `doWhenOpened((_) => Future.value()))` and then close that second database. Then you can use `ATTACH DATABASE` in the main database without having to use the migrator manually.

> Is there anyway to check the if the backup database is already in use?

You can use `customSelect('pragma database_list')` and see if the `name` column in any row contains the backup. But really, I would detach the backup as soon as you no longer need it.

**Reply by mytc** [reply]
tbh, I am quite new at this and I am not quite understand what are you trying to explain.
Let me do more digging. 
Anyway, thanks.

**Reply by simolus3** [reply]
Maybe the documentation for [attach](https://www.sqlite.org/lang_attach.html) (or similar resources) could help here.

Basically, you an attach another file to a sqlite3 database under a prefix. This allows us to copy data between databases by referencing tables in the backup file. If you have questions on anything specific I can try to answer those as well,

**Reply by mytc** [reply]
ok, I now understand the attach feature. 
Problem now is how to open another database instance? And how to use the `doWhenOpened` ?
Right now I am doing this, but don't know if this is correct way to open the database. 
```
extension CreateBackup on GeneratedDatabase {
   Future<void> myBackup(File to) async {
    try {
      var ndb = _openBackupDB(to);//correct way to open database instance?
      doWhenOpened((_) => Future.value());//how to create table here?
  
      ndb.close();
    } catch (e) {
      print(e);
    }
  }

  LazyDatabase _openBackupDB(File to) {
    return LazyDatabase(() async {
      return VmDatabase(to, logStatements: true);
    });
  }
}
```

**Reply by simolus3** [reply]
Sorry for being unclear, I think the easiest way is to open another instance of your database class. When you use that instance, it will open the connection and create all tables, which is what we want here.

So something like

```dart
Future<void> createBackup(File to, List<TableInfo> tables) async {
  moorRuntimeOptions.dontWarnAboutMultipleDatabases = true;
  final secondary = YourDatabaseClass(_openBackupDB(to));
  // Opening the database to ensure all tables are created in the `to` file!
  await secondary.doWhenOpened((_) => Future.value());
  await secondary.close();
  
  await customStatement('ATTACH ... AS backup_target');
  // If you're still getting exceptions you may want to run this in a transaction
  for (final table in tables) {
    final sourceName = table.actualTableName;
    final targetName = 'backup_target.${table.actualTableName}';

    await customStatement(
        'INSERT INTO $targetName SELECT * FROM $sourceName');
   }
  await customStatement('DETACH backup_target');
}
```

**Reply by mytc** [reply]
This is how I did:
```
class BackupIns {
  static String bc_records =
      'CREATE TABLE IF NOT EXISTS bc_records (sys INTEGER, dias INTEGER, pul INTEGER,time INTEGER, uid INTEGER)';

  openBackup(File to) async {
    moorRuntimeOptions.dontWarnAboutMultipleDatabases = true;
    final secondary = AppDatabase(_openBackupDB(to));
    await secondary.doWhenOpened((_) async {
      try {
        await secondary.customStatement(bc_records);
      } catch (e) {
        print('create table error ' + e.toString());
      }
    });
    await secondary.close();
  }

  LazyDatabase _openBackupDB(File to) {
    return LazyDatabase(() async {
      return VmDatabase(to, logStatements: true);
    });
  }
}
```
But why I can't see the file in the `to.path` via file manger?
In the debug log, I can see this:
`I/flutter (20018): Moor: Sent CREATE TABLE IF NOT EXISTS bc_records (sys INTEGER, dias INTEGER, pul INTEGER, time INTEGER, uid INTEGER) with args []`

It should be successfully created right?

**Reply by simolus3** [reply]
Yes, it should be. Do you see the original database file in your file manager?

**Reply by mytc** [reply]
Yes, I can see the original file. 
I also tried the original and secondary databases in same directory, but all i can see is the Original database file. 

The only difference is that for the backup database file path was from flutter [Filepicker package](https://pub.dev/packages/file_picker), where let user choose the backup path. 
But I tried with file creation, by using the filepicker, it works. So, i don't think there is permission issue as I already handle by permission handler package.
